Add ActiveSupport::TimeZone.country_zones helper

Envek opened the pull request:
I always wondered why there is a method to retrieve all time zones for US, but no such methods for another countries until I discovered that tzdata is already knows about all countries and their timezones and tzinfo gem even have an API for that.

Live example:

```
2.3.0 :001 > puts ActiveSupport::TimeZone.country_zones('us').map(&:to_s)
(GMT-10:00) Hawaii
(GMT-09:00) Alaska
(GMT-08:00) Pacific Time (US & Canada)
(GMT-07:00) Arizona
(GMT-07:00) Mountain Time (US & Canada)
(GMT-06:00) Central Time (US & Canada)
(GMT-05:00) Eastern Time (US & Canada)
(GMT-05:00) Indiana (East)
2.3.0 :002 > puts ActiveSupport::TimeZone.country_zones('ru').map(&:to_s)
(GMT+02:00) Kaliningrad
(GMT+03:00) Moscow
(GMT+03:00) Volgograd
(GMT+04:00) Samara
(GMT+05:00) Ekaterinburg
(GMT+06:00) Novosibirsk
(GMT+07:00) Krasnoyarsk
(GMT+08:00) Irkutsk
(GMT+09:00) Yakutsk
(GMT+10:00) Magadan
(GMT+10:00) Vladivostok
(GMT+11:00) Srednekolymsk
(GMT+12:00) Kamchatka
2.3.0 :003 > puts ActiveSupport::TimeZone.country_zones('fr').map(&:to_s)
(GMT+01:00) Paris
```

UPDATE: Method `ActiveSupport::TimeZone.us_zones` reimplemented with `ActiveSupport::TimeZone.country_zones('us')`

P.S> This PR is a good addition to my [talk](http://devconf.ru/ru/offers/offer/40) about working with timezones in Rails apps at [DevConf](http://devconf.ru/ru)

Envek committed:
Add ActiveSupport::TimeZone.country_zones helper

That helper will return time zones for any country that tzdata knows about.
So it will be much simpler for non-US people to list own country time zones
in HTML selects or anywhere.

senny commented:
@Envek can `us_zones` be implemented in terms of this new method?

Envek commented:
@senny `us_zones` can be implemented, but the order of returned time zone objects in that case is different.

Envek commented:
I added soritng to zones array, so order of zoned from `country_zones(:us)` is matching these from `us_zones` and implemented `us_zones` with this new method.

senny commented on activesupport/lib/active_support/values/time_zone.rb:
```diff
@@ -248,7 +249,19 @@ def [](arg)
       # for time zones in the USA.
       def us_zones
-        @us_zones ||= all.find_all { |z| z.name =~ /US|Arizona|Indiana|Hawaii|Alaska/ }
+        country_zones('US')
```
tests are showing the use of a symbol. Can we use `:us`?

Envek replied:
Sure we can. Anyway it will be converted to string and upcased. Will push fix in a minute.

senny commented:
:+1: /cc @rafaelfranca

Envek commented:
Can this PR be reviewed further? It's got stuck and I do not know why.

senny commented:
Looks good to me. I assigned @pixeltrix to have another look.

jonatack commented:
Nice idea @Envek.

pixeltrix commented on activesupport/lib/active_support/values/time_zone.rb:
```diff
@@ -248,7 +249,19 @@ def [](arg)
+            MAPPING.key(tz_id)
+          end.map do |tz_id|
+            self[MAPPING.key(tz_id)]
+          end.sort
```
This block is a little unwieldy - try breaking it out into a couple of private methods. Also try to do it in one map if possible so that we're not generating too many arrays. If you can't don't worry as they are cached so they're not regenerated on every request.

pixeltrix commented:
I'm :+1: on the idea - just need to fix the merge conflicts and tidy up the core loop

Envek commented:
I've simplified block that searches for ActiveSupport's `TimeZone` objects: converted `select` and `map` to `map` and `flatten` and replaced `sort` with `sort!` (unfortunately `compact!` not always returns `self` and I can not modify output of `TZInfo::Country.get(code).zone_identifiers` in place as it is frozen array). 

Now its task should be clear.

Also rebased on top of current master.

arthurnn commented on activesupport/lib/active_support/values/time_zone.rb:
```diff
@@ -248,7 +249,18 @@ def [](arg)
+      def country_zones(country_code)
+        code = country_code.to_s.upcase
+        @country_zones[code] ||=
+          TZInfo::Country.get(code).zone_identifiers.map do |tz_id|
```
will this return a `ThreadSafe::Cache` ? If not, we need to use a each here instead of a map, and push things to our `ThreadSafe::Cache` instead.

Envek replied:
Why should it return a `ThreadSafe::Cache`? I've initialized cache instance earlier and only adding key/value pairs to it, never overwriting. Key is the country code, value is the array of timezone objects.
The another cache (`@lazy_zones_map`) in this file used similarly — created beforehand and added timezone objects by timezone names as keys afterwards when needed.

thedarkone replied:
@arthurnn: I think @Envek is using `@country_zones` (`ThreadSafe::Cache`) correctly here.

PS:
> I've initialized cache instance earlier and only adding key/value pairs to it, never overwriting.

@Envek just wanted to clarify, it is thread-safe to do something like this:
```ruby
cache = ThreadSafe::Cache.new
cache[:foo] = [1,2]
cache[:foo] = [1,2,3] # rewriting/updating cache is ok
cache[:foo] += [4] # this is just syntactic sugar for cache[:foo] = cache[:foo] + [4]
```

however this wouldn't be thread-safe:
```ruby
cache = ThreadSafe::Cache.new
cache[:foo] = [1,2]
cache[:foo] << 3 # updating ruby Arrays is not thread-safe
```

PS n. 2: `ThreadSafe::Cache` has been merged into https://github.com/ruby-concurrency/concurrent-ruby, so on rebase you'll need to replace `ThreadSafe::Cache.new` with `Concurrent::Map.new`.

Envek replied:
@thedarkone thanks for the clarification! I'm only adding new key/value pairs to the cache, so no place for such a mistake, but I will remember for future use. Sadly, I can't see good docs with explanations for this neither in `thread_safe` nor in `concurrent-ruby`.
Sure, will change to `Concurrent::Map` as soon as core team will be satisfied and rebase time will come.

Envek commented:
@arthurnn @pixeltrix is there any issues left? If not I will be glad to rebase.

Envek commented:
Rebased on top of current master, changed cache storage from `ThreadSafe::Cache` to `Concurrent::Map`.

Envek commented:
@pixeltrix could you please review this PR?

Envek commented:
Is there anything prevents this PR from being merged?

P.S> Now with signed commit!
